pilot/pkg/xds/ecds.go

```diff
@@ -59,17 +69,169 @@ func (e *EcdsGenerator) Generate(proxy *model.Proxy, push *model.PushContext, w
+		secrets = e.GeneratePullSecrets(proxy, updatedSecrets, secretResources, secretController, req)
+	}
+	resources := make(model.Resources, 0, len(ec)+len(secrets))
 	for _, c := range ec {
 		resources = append(resources, &discovery.Resource{
 			Name:     c.Name,
 			Resource: util.MessageToAny(c),
 		})
 	}
+	resources = append(resources, secrets...)
+
 	return resources, model.DefaultXdsLogDetails, nil
 }
+
+func (e *EcdsGenerator) GeneratePullSecrets(proxy *model.Proxy, updatedSecrets map[model.ConfigKey]struct{}, secretResources []SecretResource,
+	secretController credscontroller.Controller, req *model.PushRequest) model.Resources {
+	if proxy.VerifiedIdentity == nil {
+		log.Warnf("proxy %s is not authorized to receive secret. Ensure you are connecting over TLS port and are authenticated.", proxy.ID)
+		return nil
+	}
+
+	results := model.Resources{}
+	for _, sr := range secretResources {
+		if updatedSecrets != nil {
+			if _, found := updatedSecrets[model.ConfigKey{Kind: gvk.Secret, Name: sr.Name, Namespace: sr.Namespace}]; !found {
+				// This is an incremental update, filter out credscontroller that are not updated.
+				continue
+			}
+		}
+
+		cachedItem, f := e.secretCache.Get(sr)
+		if f && !features.EnableUnsafeAssertions {
+			// If it is in the Cache, add it and continue
+			// We skip cache if assertions are enabled, so that the cache will assert our eviction logic is correct
+			results = append(results, cachedItem)
+			continue
+		}
+
+		cred, err := getDockerCredential(sr.Name, sr.Namespace, secretController)
+		if err != nil {
+			log.Warnf("Failed to fetch docker credential %s: %v", sr.ResourceName, err)
+		} else {
+			res := toWasmSecret(sr.ResourceName, cred)
+			results = append(results, res)
+			e.secretCache.Add(sr, req, res)
+		}
+	}
+	return results
+}
+
+func (e *EcdsGenerator) SetCredController(creds credscontroller.MulticlusterController, cache model.XdsCache) {
+	e.secretController = creds
+	e.secretCache = cache
+}
+
+func referencedSecrets(proxy *model.Proxy, push *model.PushContext, resourceNames []string) []SecretResource {
+	// The requirement for the Wasm pull secret:
+	// * Wasm pull secrets must be of type `kubernetes.io/dockerconfigjson`.
+	// * Secret are referenced by a WasmPlugin which applies to this proxy.
+	// TODO: we get the WasmPlugins here to get the secrets reference in order to decide whether ECDS push is needed,
+	//       and we will get it again at extension config build. Avoid getting it twice if this becomes a problem.
+	watched := sets.NewSet(resourceNames...)
+	wasmPlugins := push.WasmPlugins(proxy)
+	referencedSecrets := sets.Set{}
+	for _, wps := range wasmPlugins {
+		for _, wp := range wps {
+			if watched.Contains(wp.ExtensionConfiguration.Name) && wp.ImagePullSecret != "" {
+				referencedSecrets.Insert(wp.ImagePullSecret)
+			}
+		}
+	}
+	var filtered []SecretResource
+	for rn := range referencedSecrets {
+		sr, err := parseSecretName(rn, proxy.Metadata.ClusterID)
```

Comment:
So the spec says `Name of a K8s Secret in the same namespace as the WasmPlugin that` but here we expect `kubernetes://secret-namespace/secret-name`. This seems problematic that they can request any namespace

Reply:
If you look at `toSecretResourceName` implementation, it update the secret resource name to make sure it uses plugin's namespace.